doc/opimport.1.in
.TH OPIMPORT 1 "@DATE@" "oprofile @VERSION@"
.UC 4
.SH NAME
opimport \- converts sample database files
.SH SYNOPSIS
.br
.B opimport
[
.I options
]
input_file
.SH DESCRIPTION

.B opimport
converts sample database files from a foreign binary format (abi) to the native format.

.SH OPTIONS
.TP
.BI "--abi / -a"
Input abi file description location.
.br
.TP
.BI "--force / -f"
Force conversion even if the input and output abi are identical.
.br
.TP
.BI "--output / -o filename"
Specify the output filename. If the output file already exists it is not overwritten but data are accumulated in. Sample filename are informative
for post profile tools and must be kept identical, in other word the pathname
from the first path component containing a '{' must be kept as it in the
output filename.
.br
.TP
.BI "--help / -? / --usage"
Show help message.
.br
.TP
.BI "--verbose / -V"
Give verbose debugging output.
.br
.TP
.BI "--version / -v"
Show version.

.SH ENVIRONMENT
No special environment variables are recognized by opimport

.SH FILES
.TP
.I <session_dir>/abi
The abi file description of the sample database files

.SH VERSION
.TP
This man page is current for @PACKAGE@-@VERSION@.

.SH SEE ALSO
.BR @OP_DOCDIR@,
.BR oprofile(1)


ChangeLog
2011-02-09  William Cohen  <wcohen@redhat.com>

	* libpp/format_output.h: Do not use mutable for reference variable.

2011-01-17  Will Deacon  <will.deacon@arm.com>

	* events/arm/armv7-ca9/events: Add missing TLB event

2011-01-05  William Cohen  <wcohen@redhat.com>

	* utils/opcontrol: Add argument checking for numerical arguments

2010-12-16  John Villalovos <john.l.villalovos@intel.com>

	* events/Makefile.am:
	* libop/op_cpu_type.c:
	* libop/op_cpu_type.h:
	* libop/op_hw_specific.h:
	* libop/op_events.c:
	* utils/ophelp.c:
	* events/i386/westmere/events (new):
	* events/i386/westmere/unit_masks (new): Add support for Intel
	  Westmere micro-architecture processors

2010-12-15  Will Cohen  <wcohen@redhat.com>

	* libop/op_cpu_type.c:
	* libop/op_cpu_type.h:
	* libop/op_hw_specific.h: User-space identification of processors
	  that support Intel architectural events

2010-12-14  Suravee Suthikulpanit <suravee.suthikulpanit@amd.com>

	* oprofile/daemon/opd_ibs_trans.c: Fix non-x86 build issue 
	  due to cpuid instruction

2010-11-7  Suravee Suthikulpanit <suravee.suthikulpanit@amd.com>

	* oprofile/utils/opcontrol:
	* oprofile/daemon/init.c:
	* oprofile/daemon/opd_extended.h:
	* oprofile/daemon/opd_extended.c:
	* oprofile/daemon/opd_ibs.h:
	* oprofile/daemon/opd_ibs.c:
	* oprofile/daemon/opd_ibs_trans.h:
	* oprofile/daemon/opd_ibs_trans.c:
	* oprofile/events/x86-64/family10/unit_masks:
	* oprofile/daemon/opd_ibs_macro.h: Additional IBS supports for 
	  family12/14/15h including:
		- IBS Op branch target address log
		- IBS Op memory access log
		- IBS Op extended count bits

2010-11-7  Suravee Suthikulpanit <suravee.suthikulpanit@amd.com>

	* events/Makefile.am:
	* libop/op_events.c:
	* libop/op_cpu_type.h:
	* libop/op_cpu_type.c:
	* utils/ophelp.c:
	* events/x86-64/family12h/events: New File
	* events/x86-64/family12h/unit_masks: New File
	* events/x86-64/family14h/events: New File
	* events/x86-64/family14h/unit_masks: New File
	* events/x86-64/family15h/events: New File
	* events/x86-64/family15h/unit_masks: New File
	  Add support for new AMD processors (family12h/14h/15h)

2010-10-15  Roland Grunberg  <roland.grunberg@gmail.com>

	* libop/op_xml_events.c:
	* libop/op_xml_out.c:
	* libop/op_xml_out.h:
	* doc/ophelp.xsd: Add unit mask type attribute for an event in
	  the ophelp schema

2010-10-15  Maynard Johnson  <maynardj@us.ibm.com>

	* doc/ophelp.xsd:
	* libop/op_xml_events.c: Fix schema validation issues and error in
	  xml generation

2010-10-13  Maynard Johnson  <maynardj@us.ibm.com>

	* libabi/opimport.cpp: Fix uninitialized variable warning when
	  building with gcc 4.4

2010-10-13  Maynard Johnson  <maynardj@us.ibm.com>

	* events/mips/Makefile.am: Correction to 8/26 MIPS patch
	  to add events and unit_masks to makefile

2010-10-07  William Cohen  <wcohen@redhat.com>

	* events/i386/arch_perfmon/events: Correct filter values.

2010-08-26  Paul Lind  <plind@mips.com>

	* libop/op_cpu_type.[h,c]:
	* libop/op_events.c:
	* utils/ophelp.c: Add "mips/74K" and "mips/1004K" as valid cpu types,
	  and change number of available 34K counters to 2.
	* libop/tests/alloc_counter_tests.c: correct 34K test to 2 counters.
	* events/mips/24K/events: correct the performance counters for 24K.
	* events/mips/34K/events: correct the performance counters for 34K.
	* events/mips/74K/[events,unit_masks]: Add events for 74K.
	* events/mips/1004K/[events,unit_masks]: Add events for 1004K.

2010-08-13  Maynard Johnson  <maynardj@us.ibm.com>

	* utils/opcontrol:
	* doc/opcontrol.1.in: Update help and man page for opcontrol to indicate
	  that buffer values may be reset to default values by passing a '0'

2010-08-02  Maynard Johnson  <maynardj@us.ibm.com>

	* utils/opcontrol:
	* libpp/profile_spec.cpp:
	* pp/oparchive.cpp:  Moved the copying of stats to opcontrol::do_dump_data
	  and removed the unnecessary and confusing message that indicated
	  when overflow stats were not available.

2010-07-13  Maynard Johnson  <maynardj@us.ibm.com>

	* events/arm/armv7-common/events: (new)
	* events/arm/armv7-common/unit_masks: (new)
	* events/arm/armv7-ca9/events: (new)
	* events/arm/armv7-ca9/unit_masks: (new)
	  I neglected to do 'cvs add' for these new files in the
	  previous commit; doing so now.

2010-07-01  Will Deacon  <will.deacon@arm.com>

	* events/arm/armv7-common/events:
	* events/arm/armv7-common/unit_masks:
	* events/arm/armv7/events: factor out ARMv7 common architectural events
	* events/Makefile.am:
	* libop/op_cpu_type.c:
	* libop/op_cpu_type.h:
	* libop/op_events.c:
	* utils/ophelp.c: correct usage of core terminology for v7 and MPCore
	* events/arm/armv7-ca9/events:
	* events/arm/armv7-ca9/unit_masks: add support for Cortex-A9 events

2010-06-21  John Villalovos <john.l.villalovos@intel.com>

	* libop/op_cpu_type.[h,c]:
	* libop/op_events.c:
	* utils/ophelp.c: Add "i386/nehalem" as a valid cpu_type.  This will
	allow the kernel in the future (if desired) to use "i386/nehalem"
	instead of "i386/core_i7".  Core i7 processors can be either Nehalem
	or Westmere microarchitecture but the "i386/core_i7" cpu_type in
	Oprofile is only for Nehalem microarchitecture processors.

2010-06-11  William Cohen <wcohen@redhat.com>

        * libregex/stl.pat.in: Avoid machine specific configuration.

2010-05-18  Daniel Hansel  <daniel.hansel@linux.vnet.ibm.com>

	* doc/oprofile.xml: Document that only kernel versions 2.6.13 or
	  later provide support for anonymous mapped regions

2010-04-13  Maynard Johnson  <maynardj@us.ibm.com>

	* libutil++/bfd_support.cpp: Fix up translate_debuginfo_syms
	  so it doesn't rely on section index values being the same
	  between real image and debuginfo file (to resolve problem
	  reported by Will Cohen on Fedora 12)

2010-03-25  Oliver Schneider  <oliver@assarbad.net>

	* libpp/parse_filename.cpp:  Catch case where a basic_string::erase
	  error can occur in opreport when parsing an invalid sample file name

2010-03-25  Maynard Johnson  <maynardj@us.ibm.com>


libabi/opimport.cpp
/**
 * @file opimport.cpp
 * Import sample files from other ABI
 *
 * @remark Copyright 2002 OProfile authors
 * @remark Read the file COPYING
 *
 * @author Graydon Hoare
 */

#include "abi.h"
#include "odb.h"
#include "popt_options.h"
#include "op_sample_file.h"

#include <fstream>
#include <iostream>
#include <sstream>
#include <vector>
#include <cassert>
#include <cstring>
#include <cstdlib>

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/mman.h>
#include <cstdlib>
#include <cstring>

using namespace std;

namespace {
	string output_filename;
	string abi_filename;
	bool verbose;
	bool force;
};


popt::option options_array[] = {
	popt::option(verbose, "verbose", 'V', "verbose output"),
	popt::option(output_filename, "output", 'o', "output to file", "filename"),
	popt::option(abi_filename, "abi", 'a', "abi description", "filename"),
	popt::option(force, "force", 'f', "force conversion, even if identical")
};


struct extractor {

	abi const & theabi;

	unsigned char const * begin;
	unsigned char const * end;
	bool little_endian;

	explicit
	extractor(abi const & a, unsigned char const * src, size_t len)
		: theabi(a), begin(src), end(src + len) {
		little_endian = theabi.need(string("little_endian")) == 1;
		if (verbose) {
			cerr << "source byte order is: "
			     << string(little_endian ? "little" : "big")
			     << " endian" << endl;
		}
	}

	template <typename T>
	void extract(T & targ, void const * src_,
	             char const * sz, char const * off);
};


template <typename T>
void extractor::extract(T & targ, void const * src_,
                        char const * sz, char const * off)
{
	unsigned char const * src = static_cast<unsigned char const *>(src_)
		+ theabi.need(off);
	size_t nbytes = theabi.need(sz);

	targ = 0;
	if (nbytes == 0)
		return;
	
	assert(nbytes <= sizeof(T));
	assert(src >= begin);
	assert(src + nbytes <= end);
	
	if (verbose) {
		ostringstream message;
		message << hex << "get " << sz << " = " << nbytes
		        << " bytes @ " << off << " = " << (src - begin)
		        << " : ";
		cerr << message.str();
	}

	if (little_endian)
		while(nbytes--)
			targ = (targ << 8) | src[nbytes];
	else
		for(size_t i = 0; i < nbytes; ++i)
			targ = (targ << 8) | src[i];
	
	if (verbose)
		cerr << " = " << targ << endl;
}


void import_from_abi(abi const & abi, void const * srcv,
                     size_t len, odb_t * dest) throw (abi_exception)
{
	struct opd_header * head =
		static_cast<opd_header *>(odb_get_data(dest));
	unsigned char const * src = static_cast<unsigned char const *>(srcv);
	unsigned char const * const begin = src;
	extractor ext(abi, src, len);	

	memcpy(head->magic, src + abi.need("offsetof_header_magic"), 4);

	// begin extracting opd header
	ext.extract(head->version, src, "sizeof_u32", "offsetof_header_version");
	ext.extract(head->cpu_type, src, "sizeof_u32", "offsetof_header_cpu_type");
	ext.extract(head->ctr_event, src, "sizeof_u32", "offsetof_header_ctr_event");
	ext.extract(head->ctr_um, src, "sizeof_u32", "offsetof_header_ctr_um");
	ext.extract(head->ctr_count, src, "sizeof_u32", "offsetof_header_ctr_count");
	ext.extract(head->is_kernel, src, "sizeof_u32", "offsetof_header_is_kernel");
	// "double" extraction is unlikely to work
	head->cpu_speed = 0.0;
	ext.extract(head->mtime, src, "sizeof_u64", "offsetof_header_mtime");
	ext.extract(head->cg_to_is_kernel, src, "sizeof_u32",
		"offsetof_header_cg_to_is_kernel");
	ext.extract(head->anon_start, src, "sizeof_u32",
		"offsetof_header_anon_start");
	ext.extract(head->cg_to_anon_start, src, "sizeof_u32",
		"offsetof_header_cg_to_anon_start");
	src += abi.need("sizeof_struct_opd_header");
	// done extracting opd header

	// begin extracting necessary parts of descr
	odb_node_nr_t node_nr;
	ext.extract(node_nr, src, "sizeof_odb_node_nr_t", "offsetof_descr_current_size");
	src += abi.need("sizeof_odb_descr_t");
	// done extracting descr

	// skip node zero, it is reserved and contains nothing usefull
	src += abi.need("sizeof_odb_node_t");

	// begin extracting nodes
	unsigned int step = abi.need("sizeof_odb_node_t");
	if (verbose)
		cerr << "extracting " << node_nr << " nodes of " << step << " bytes each " << endl;

	assert(src + (node_nr * step) <= begin + len);

	for (odb_node_nr_t i = 1 ; i < node_nr ; ++i, src += step) {
		odb_key_t key;
		odb_value_t val;
		ext.extract(key, src, "sizeof_odb_key_t", "offsetof_node_key");
		ext.extract(val, src, "sizeof_odb_value_t", "offsetof_node_value");
		int rc = odb_add_node(dest, key, val);
		if (rc != EXIT_SUCCESS) {
			cerr << strerror(rc) << endl;
			exit(EXIT_FAILURE);
		}
	}
	// done extracting nodes
}


int main(int argc, char const ** argv)
{

	vector<string> inputs;
	popt::parse_options(argc, argv, inputs);

	if (inputs.size() != 1) {
		cerr << "error: must specify exactly 1 input file" << endl;
		exit(1);
	}

	if (inputs[0].rfind(".jo") == inputs[0].size() - 3) {
		if (verbose) {
			cerr << "Found a .jo file. Import is not allowed or necessary. Done." << endl;
			cerr << inputs[0] << endl;
		}
		// Silently exit with success.
		exit(EXIT_SUCCESS);
	}

	abi current_abi, input_abi;

	{
		ifstream abi_file(abi_filename.c_str());
		if (!abi_file) {
			cerr << "error: cannot open abi file "
			     << abi_filename << endl;
			exit(1);
		}
		abi_file >> input_abi;
	}

	if (!force && current_abi == input_abi) {
		cerr << "input abi is identical to native. "
		     << "no conversion necessary." << endl;
		exit(1);
	}

	int in_fd;
	struct stat statb;
	void * in;
	odb_t dest;
	int rc;

	in_fd = open(inputs[0].c_str(), O_RDONLY);
	assert(in_fd > 0);
	assert(fstat(in_fd, &statb) == 0);
	in = mmap(0, statb.st_size, PROT_READ, MAP_PRIVATE, in_fd, 0);
	assert(in != (void *)-1);

	rc = odb_open(&dest, output_filename.c_str(), ODB_RDWR,
		      sizeof(struct opd_header));
	if (rc) {
		cerr << "odb_open() fail:\n"
		     << strerror(rc) << endl;
		exit(EXIT_FAILURE);
	}

	try {
		import_from_abi(input_abi, in, statb.st_size, &dest);
	} catch (abi_exception & e) {
		cerr << "caught abi exception: " << e.desc << endl;
	}

	odb_close(&dest);

	assert(munmap(in, statb.st_size) == 0);
}
